accord-core/src/main/java/accord/primitives/Keys.java

```diff
@@ -0,0 +1,449 @@
+    }
+
+    public int findNext(Key key, int startIndex)
+    {
+        return SortedArrays.exponentialSearch(keys, startIndex, keys.length, key);
+    }
+
+    // returns thisIdx in top 32 bits, thatIdx in bottom
+    public long findNextIntersection(int thisIdx, Keys that, int thatIdx)
+    {
+        return SortedArrays.findNextIntersection(this.keys, thisIdx, that.keys, thatIdx);
+    }
+
+    public Keys with(Key key)
+    {
+        int insertPos = Arrays.binarySearch(keys, key);
+        if (insertPos >= 0)
+            return this;
+        insertPos = -1 - insertPos;
+
+        Key[] src = keys;
+        Key[] trg = new Key[src.length + 1];
+        System.arraycopy(src, 0, trg, 0, insertPos);
+        trg[insertPos] = key;
+        System.arraycopy(src, insertPos, trg, insertPos + 1, src.length - insertPos);
+        return new Keys(trg);
+    }
+
+    public Stream<Key> stream()
+    {
+        return Stream.of(keys);
+    }
+
+    @Override
+    public Iterator<Key> iterator()
+    {
+        return new Iterator<>()
+        {
+            int i = 0;
+            @Override
+            public boolean hasNext()
+            {
+                return i < keys.length;
+            }
+
+            @Override
+            public Key next()
+            {
+                return keys[i++];
+            }
+        };
+    }
+
+    public static Keys of(Key k0, Key... kn)
+    {
+        Key[] keys = new Key[kn.length + 1];
+        keys[0] = k0;
+        for (int i=0; i<kn.length; i++)
+            keys[i + 1] = kn[i];
+
+        return new Keys(keys);
+    }
+
+    public boolean any(KeyRanges ranges, Predicate<Key> predicate)
+    {
+        return 1 == foldl(ranges, (i1, key, i2, i3) -> predicate.test(key) ? 1 : 0, 0, 0, 1);
+    }
+
+    public boolean any(IndexedPredicate<Key> predicate)
+    {
+        return 1 == foldl((i, key, p, v) -> predicate.test(i, key) ? 1 : 0, 0, 0, 1);
+    }
+
+    /**
+     * Count the number of keys matching the predicate and intersecting with the given ranges.
+     * If terminateAfter is greater than 0, the method will return once terminateAfter matches are encountered
+     */
+    @Inline
+    public <V> V foldl(KeyRanges rs, IndexedFold<Key, V> fold, V accumulator)
+    {
+        int ai = 0, ri = 0;
+        while (true)
+        {
+            long ari = rs.findNextIntersection(ri, this, ai);
+            if (ari < 0)
+                break;
+
+            ai = (int)(ari >>> 32);
+            ri = (int)ari;
+            KeyRange range = rs.get(ri);
+            int nextai = range.higherKeyIndex(this, ai + 1, keys.length);
+            while (ai < nextai)
+            {
+                accumulator = fold.apply(ai, get(ai), accumulator);
+                ++ai;
+            }
+        }
+
+        return accumulator;
+    }
+
+    @Inline
+    public long foldl(KeyRanges rs, IndexedFoldToLong<Key> fold, long param, long initialValue, long terminalValue)
```

Comment:
feel we can get rid of `long param` as its just there to pass a `shards.length` to `ShardedRanges::addKeyIndex`, every other call path has access to it already.

Reply:
If we get rid of it then `ShardedRanges::addKeyIndex` likely becomes a non-static lambda.